membership/cb_membership.go

```diff
@@ -0,0 +1,261 @@
+package membership
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	jsoniter "github.com/json-iterator/go"
+
+	"github.com/Trendyol/go-dcp-client/dcp"
+
+	"github.com/google/uuid"
+
+	"github.com/Trendyol/go-dcp-client/helpers"
+	"github.com/Trendyol/go-dcp-client/logger"
+	"github.com/Trendyol/go-dcp-client/membership/info"
+	"github.com/couchbase/gocbcore/v10"
+	"github.com/couchbase/gocbcore/v10/memd"
+)
+
+type cbMembership struct {
+	info                *info.Model
+	infoChan            chan *info.Model
+	client              dcp.Client
+	id                  []byte
+	clusterJoinTime     int64
+	lastActiveInstances []Instance
+	handler             info.Handler
+	monitorQuery        []byte
+	indexQuery          []byte
+}
+
+type Instance struct {
+	ID              *string `json:"id,omitempty"`
+	Type            string  `json:"type"`
+	HeartbeatTime   int64   `json:"heartbeatTime"`
+	ClusterJoinTime int64   `json:"clusterJoinTime"`
+}
+
+const (
+	_type                  = "instance"
+	_expirySec             = 2
+	_heartbeatIntervalSec  = 1
+	_heartbeatToleranceSec = 2
+	_monitorIntervalMs     = 500
+	_timeoutSec            = 10
+)
+
+func (h *cbMembership) GetInfo() *info.Model {
+	if h.info != nil {
+		return h.info
+	}
+
+	return <-h.infoChan
+}
+
+func (h *cbMembership) register() {
+	ctx, cancel := context.WithTimeout(context.Background(), _timeoutSec*time.Second)
+	defer cancel()
+
+	now := time.Now().UnixNano()
+
+	h.clusterJoinTime = now
+
+	instance := Instance{
+		Type:            _type,
+		HeartbeatTime:   now,
+		ClusterJoinTime: now,
+	}
+
+	err := h.client.UpdateDocument(ctx, h.id, instance, _expirySec)
+
+	var kvErr *gocbcore.KeyValueError
+	if err != nil && errors.As(err, &kvErr) && kvErr.StatusCode == memd.StatusKeyNotFound {
+		err = h.client.CreateDocument(ctx, h.id, instance, _expirySec)
+
+		if err == nil {
+			err = h.client.UpdateDocument(ctx, h.id, instance, _expirySec)
+		}
+	}
+
+	if err != nil {
+		logger.Panic(err, "error while register")
+		return
+	}
+}
+
+func (h *cbMembership) isClusterChanged(currentActiveInstances []Instance) bool {
+	if len(h.lastActiveInstances) != len(currentActiveInstances) {
```

Comment:
You can discard this if block, we just need to compare last active instances' ids are same with the current active instances' ids. Count comparison can lead to wrong results. Imagine a scenario like following:
your cluster like this at time t:
instance with id 1
instance with id 2
instance with id 3
alive instance count is 3

then suddenly instance with id 2 is crashed at time t + 50 ms
instance with id 1
~~instance with id 2~~
instance with id 3
alive instance count is 2

then imagine instance with id 2 is restarted and generated new id 4 and now your cluster looks like this at time t + 100 ms:
instance with id 1
instance with id 3
instance with id 4
alive instance count is 3

And you monitored again the cluster at the time t + 500 ms let's say. You still see instance count as 3 but the cluster is changed, clusterJoinTime ordering is changed.